Source/Atomic/Network/MasterServerClient.cpp

```diff
@@ -0,0 +1,556 @@
+            udpSecondsTillRetry_ = 0.5;
+        }
+        else
+        {
+            udpSecondsTillRetry_ -= dt;
+            udpConnectionSecondsRemaining_ -= dt;
+        }
+    }
+
+}
+
+void MasterServerClient::SetConnectToMasterState(ConnectToMasterState state)
+{
+    Atomic::Network* network = GetSubsystem<Network>();
+    kNet::Network* kNetNetwork = network->GetKnetNetwork();
+
+    if (connectToMasterState_ == MASTER_NOT_CONNECTED &&
+        state == MASTER_CONNECTING_TCP)
+    {
+        masterTCPConnection_ = kNetNetwork->ConnectSocket(masterServerInfo_.address.CString(),
+                                                          masterServerInfo_.port, kNet::SocketOverTCP);
+
+        if (!masterTCPConnection_)
+        {
+            SetConnectToMasterState(MASTER_CONNECTION_FAILED);
+        }
+        else
+        {
+            SendMessageToMasterServer("{ \"cmd\": \"connectRequest\" }");
+        }
+    }
+    else if (connectToMasterState_ == MASTER_CONNECTING_TCP &&
+             state == MASTER_CONNECTING_UDP)
+    {
+        Atomic::Network* network = GetSubsystem<Network>();
+        kNet::Network* kNetNetwork = network->GetKnetNetwork();
+
+        kNet::NetworkServer* server = kNetNetwork->GetServer().ptr();
+
+        kNet::EndPoint masterEndPoint;
+
+        sscanf(masterServerInfo_.address.CString(), "%hu.%hu.%hu.%hu",
+               (unsigned short *) &masterEndPoint.ip[0],
+               (unsigned short *) &masterEndPoint.ip[1],
+               (unsigned short *) &masterEndPoint.ip[2],
+               (unsigned short *) &masterEndPoint.ip[3]);
+
+        masterEndPoint.port = masterServerInfo_.port;
+
+
+        if (server)
+        {
+            std::vector < kNet::Socket * > listenSockets = kNetNetwork->GetServer()->ListenSockets();
+
+            kNet::Socket *listenSocket = listenSockets[0];
+
+            // Create a UDP and a TCP connection to the master server
+            // UDP connection re-uses the same udp port we are listening on for the sever
+            masterUDPConnection_ = new kNet::Socket(listenSocket->GetSocketHandle(),
+                                                    listenSocket->LocalEndPoint(),
+                                                    listenSocket->LocalAddress(), masterEndPoint, "",
+                                                    kNet::SocketOverUDP, kNet::ServerClientSocket, 1400);
+        }
+        else
+        {
+            masterUDPConnection_ = kNetNetwork->CreateUnconnectedUDPSocket(masterServerInfo_.address.CString(),
+                                                                           masterServerInfo_.port);
+        }
+    }
+
+    if (state == MASTER_CONNECTION_FAILED)
+    {
+        LOGERROR("Could not connect to master server");
+        SendEvent(E_MASTERCONNECTIONFAILED);
+    }
+
+    if (state == MASTER_CONNECTED)
+    {
+        SendEvent(E_MASTERCONNECTIONREADY);
+    }
+
+    connectToMasterState_ = state;
```

Comment:
Great work, I am pleasantly surprised to find a full master server implementation nestled in this PR :+1: